File: packages/sdk-node/src/integrations/anthropic.ts
import type { AuditTrailClient, ActiveSpan } from '../client';
import { completionEvent, promptEvent } from '../events';
import { redactSecrets } from '../redact';

export interface AnthropicMessageParams {
  model: string;
  max_tokens: number;
  messages: Array<{ role: string; content: string | unknown }>;
  [key: string]: unknown;
}

export interface AnthropicMessageResult {
  id?: string;
  model?: string;
  content?: Array<{ type: string; text?: string }>;
  usage?: { input_tokens?: number; output_tokens?: number };
  [key: string]: unknown;
}

export interface AnthropicClientLike {
  messages: {
    create(params: AnthropicMessageParams): Promise<AnthropicMessageResult>;
  };
}

/**
 * Wrap Anthropic messages.create and emit prompt/completion audit events.
 */
export function wrapAnthropicMessages(
  client: AuditTrailClient,
  anthropic: AnthropicClientLike,
  span: ActiveSpan,
): AnthropicClientLike {
  return {
    messages: {
      create: async (params) => {
        span.addEvent(promptEvent(span.spanId, redactSecrets({ ...params })));
        const result = await anthropic.messages.create(params);
        span.addEvent(
          completionEvent(
            span.spanId,
            redactSecrets({
              id: result.id,
              model: result.model ?? params.model,
              content: result.content,
              usage: result.usage,
            }),
          ),
        );
        await client.flush();
        return result;
      },
    },
  };
}

File: packages/sdk-python/audit_trail/integrations/anthropic.py
from __future__ import annotations

from typing import Any, Callable

from ..client import ActiveSpan, AuditTrailClient
from ..redact import redact_secrets


def wrap_anthropic_messages(
    client: AuditTrailClient,
    create_fn: Callable[..., Any],
    span: ActiveSpan,
) -> Callable[..., Any]:
    def traced_create(*args: Any, **kwargs: Any) -> Any:
        params = kwargs if kwargs else (args[0] if args else {})
        span.add_event(
            {
                "schema_version": "1.0",
                "event_id": f"evt_{span.span_id}_prompt",
                "type": "prompt",
                "occurred_at": _utc_now(),
                "span_id": span.span_id,
                "payload": redact_secrets(params if isinstance(params, dict) else {"args": params}),
            }
        )
        result = create_fn(*args, **kwargs)
        span.add_event(
            {
                "schema_version": "1.0",
                "event_id": f"evt_{span.span_id}_completion",
                "type": "completion",
                "occurred_at": _utc_now(),
                "span_id": span.span_id,
                "payload": redact_secrets(result if isinstance(result, dict) else {"result": result}),
            }
        )
        client.flush()
        return result

    return traced_create


def _utc_now() -> str:
    import time

    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
